jkq/ddsim/hybridqasmsimulator.py

```diff
@@ -0,0 +1,170 @@
+        if isinstance(quantum_circuits, QasmQobj) or isinstance(quantum_circuits, PulseQobj):
+            raise QiskitError('QasmQobj and PulseQobj are not supported.')
+
+        if not isinstance(quantum_circuits, list):
+            quantum_circuits = [quantum_circuits]
+
+        out_options = {}
+        for key in options:
+            if not hasattr(self.options, key):
+                warnings.warn("Option %s is not used by this backend" % key, UserWarning, stacklevel=2)
+            else:
+                out_options[key] = options[key]
+        circuit_qobj = assemble(quantum_circuits, self, **out_options)
+
+        job_id = str(uuid.uuid4())
+        local_job = JKQJob(self, job_id, self._run_job, circuit_qobj, **options)
+        local_job.submit()
+        return local_job
+
+    def _run_job(self, job_id, qobj_instance: Qobj, **options):
+        self._validate(qobj_instance)
+
+        start = time.time()
+        result_list = [self.run_experiment(qobj_exp, **options) for qobj_exp in qobj_instance.experiments]
+        end = time.time()
+
+        result = {'backend_name': self.configuration().backend_name,
+                  'backend_version': self.configuration().backend_version,
+                  'qobj_id': qobj_instance.qobj_id,
+                  'job_id': job_id,
+                  'results': result_list,
+                  'status': 'COMPLETED',
+                  'success': True,
+                  'time_taken': (end - start),
+                  'header': qobj_instance.header.to_dict()
+                  }
+        return Result.from_dict(result)
+
+    def run_experiment(self, qobj_experiment: QasmQobjExperiment, **options):
+        start_time = time.time()
+        seed = options.get('seed', -1)
+        mode = options.get('mode', 'amplitude')
+        nthreads = int(options.get('nthreads', local_hardware_info()['cpus']))
+        if mode == 'amplitude':
+            hybrid_mode = ddsim.HybridMode.amplitude
+            max_qubits = int(log2(local_hardware_info()['memory'] * (1024 ** 3) / 16))
+            algorithm_qubits = qobj_experiment.header.n_qubits
+            if algorithm_qubits > max_qubits:
+                raise JKQSimulatorError('Not enough memory available to simulate the circuit even on a single thread')
+            qubit_diff = max_qubits - algorithm_qubits
+            nthreads = int(min(2 ** qubit_diff, nthreads))
+        elif mode == 'dd':
+            hybrid_mode = ddsim.HybridMode.DD
+        else:
+            raise JKQSimulatorError('Simulation mode', mode, 'not supported by JKQ hybrid simulator. Available modes are \'amplitude\' and \'dd\'')
+
+        sim = ddsim.HybridCircuitSimulator(qobj_experiment, seed, hybrid_mode, nthreads)
+
+        shots = options['shots']
+        if self.SHOW_STATE_VECTOR and shots > 0:
+            logger.info('Statevector can only be shown if shots == 0 when using the amplitude hybrid simulation mode. Setting shots=1.')
```

Comment:
Should it be `... Setting shots=0.')` at the end?

Reply:
Ups. Good catch